File: neetcode/blind75/trees/same_binary_tree.py
# https://neetcode.io/problems/same-binary-tree

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isSameTree(self, p: Optional[TreeNode], q: Optional[TreeNode]) -> bool:
        
        # recursive DFS
        # if p is None and q is None:
        #     return True
        # elif p and q and p.val == q.val:
        #     left = self.isSameTree(p.left, q.left)
        #     right = self.isSameTree(p.right, q.right)
        #     return (left and right)
        # else:
        #     return False

        # BFS
        qu_p = []
        qu_q = []

        qu_p.append(p)
        qu_q.append(q)

        while len(qu_p) > 0 and len(qu_q) > 0:
            pop_p = qu_p.pop(0)
            pop_q = qu_q.pop(0)

            if pop_p is None and pop_q is None:
                continue
            elif pop_p and pop_q and pop_p.val == pop_q.val:
                qu_p.append(pop_p.left)
                qu_q.append(pop_q.left)

                qu_p.append(pop_p.right)
                qu_q.append(pop_q.right)
            else:
                return False

        if len(qu_q) != 0 or len(qu_q) != 0:
            return False
        else:
            return True

File: neetcode/blind75_JS/Trees/same-binary-tree.js
// https://neetcode.io/problems/same-binary-tree

/**
 * Definition for a binary tree node.
 * class TreeNode {
 *     constructor(val = 0, left = null, right = null) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */

class Solution {
    /**
     * @param {TreeNode} p
     * @param {TreeNode} q
     * @return {boolean}
     */
    isSameTree(p, q) {
        console.log('lorem')
        console.log(p)
        console.log(q)
        if (p === null && q === null) {
            return true
        }

        if (p !== null && q !== null && p.val === q.val) {
            return this.isSameTree(p.left, q.left) && this.isSameTree(p.right, q.right)
        } else {
            return false
        }

        // BFS
        // if (p === null && q === null) {
        //     return true
        // } else if (p === null || q === null) {
        //     return false
        // }
        // const quP = []
        // const quQ = []
        // quP.push(p)
        // quQ.push(q)

        // while (quP.length && quQ.length) {
        //     let currP = quP.shift()
        //     let currQ = quQ.shift()

        //     if (currP === null && currQ === null) {
        //         continue
        //     }

        //     if (currP === null || currQ === null || currP.val !== currQ.val) {
        //         return false
        //     }

        //     quP.push(currP.left)
        //     quP.push(currP.right)
        //     quQ.push(currQ.left)
        //     quQ.push(currQ.right)
            
        // }

        // return true
    }
}
